src/registry/docsetregistry.cpp

```diff
@@ -121,25 +125,32 @@ void DocsetRegistry::_addDocset(const QString &path)
     emit docsetAdded(name);
 }
 
-void DocsetRegistry::search(const QString &query)
+void DocsetRegistry::search(const QString &query, CancellationToken token)
 {
-    // Only invalidate queries
-    if (query.isEmpty())
-        return;
-
-    QMetaObject::invokeMethod(this, "_runQuery", Qt::QueuedConnection, Q_ARG(QString, query));
+    qRegisterMetaType<CancellationToken>("CancellationToken");
+    qRegisterMetaType<CancellationToken>("Zeal::CancellationToken");
+    QMetaObject::invokeMethod(this, "_runQueryAsync", Qt::QueuedConnection,
+                              Q_ARG(QString, query), Q_ARG(CancellationToken, token));
 }
 
-void DocsetRegistry::_runQuery(const QString &query)
+void MergeQueryResults(QList<SearchResult> &finalResult, const QList<SearchResult> &partial)
 {
-    m_queryResults.clear();
-
-    for (Docset *docset : docsets())
-        m_queryResults << docset->search(query);
+    finalResult += partial;
+}
 
-    std::sort(m_queryResults.begin(), m_queryResults.end());
+void DocsetRegistry::_runQueryAsync(const QString &query, const CancellationToken token)
+{
+    QFuture<QList<SearchResult> > queryResultsFuture = QtConcurrent::mappedReduced(
```

Comment:
It definitely is possible to get rid of `invokeMethod`, but I'd rather do that in a separate PR.
Also, I was used to writing `> >` and only realized that this is not necessary after pushing this PR.